Mango
=====

chia pet analysis software


Citation
========

Mango.R: a bias-correcting ChIA-PET data analysis pipeline
Phanstiel DH, Boyle AP,  Snyder MP.  In Preparation. 


Mango Installation
==================

1. Mango depends on the following R packages.

 a) hash
 b) Rcpp

They can be installed throug CRAN. For example to install the package 'hash' open R and type the following

::

  install.packages('hash')

2. Mango depends on the following software pacakges which should be installed prior to using Mango.

 a) Bowtie     (http://bowtie-bio.sourceforge.net)
 b) Bedtools   (https://github.com/arq5x/bedtools2)
 c) MACS2      (https://github.com/taoliu/MACS)


3. Once dependencies are installed Mango can be installed from the command line using the following command.

::

  R CMD INSTALL Mango_0.9.9.tar.gz


Features
========

Mango uses fastq files generated by illumina sequencers to call peaks and interactions from ChIA PET experiments.  Arguments can be passed to Mango either by a configuration file, through the command line, or a combination of both.  In cases where arguments at supplied both through the command line and a configuration file the values passed via command line arguments will be used.




Usage of Mango
==============

::

  Rscript Mango.R [-options]

Example for regular interactions calling
:: 

Rscript Mango.R fastq1=samplename_1.fastq fastq2=samplename_1.fastq prefix=samplename argfile=argfile.txt chromexclude=chrM,chrY stages=1:5


Example of a argfile
:: 

bowtiepath        = /path/to/bowtie
bedtoolspath      = /path/to/bedtools
macs2path         = /path/to/macs2
bowtieref         = /path/to/hg19
bedtoolsgenome    = /path/to/human.hg19.genome


Parameters
----------


ALL STAGES
~~~~~~~~~~


stages STAGES
```````````````````````

stages of the pipeline to execute.  STAGE can be either a single stage (e.g 1 or a range of stagnes e.g 1:5). default = 1:5

prefix PREFIX
```````````````````````

prefix for all output files. default = mango


outdir OUTDIR
```````````````````````

The output direcoroy. default = NULL


chrominclude CHROMINCLUDE
```````````````````````

comma separated list of chromosomes to use (e.g. chr1,chr2,chr3,...).  Only these chromosomes will be processed.  If NULL all chromosomes with be processed. default = NULL


chromexclude CHROMECXLUDE
```````````````````````

comma separated list of chromosomes to exclude (e.g. chrM,chrY).  If NULL all chromosomes with be processed. default = NULL

STAGE 1 PARAMETERS
~~~~~~~~~~

linkerA LINKERA
```````````````````````

linker sequence to look for. default = GTTGGATAAG

linkerB LINKERB
```````````````````````

linker sequence to look for. default = GTTGGAATGT

minlength MINLENGTH
```````````````````````

min length of reads after linker trimming. default = 15

maxlength MAXLENGTH
```````````````````````

max length of reads after linker trimming. default = 25

keepempty KEEPEMPTY
```````````````````````

Should reads with no linker be kept (TRUE or FALSE). default = FALSE


STAGE 2 PARAMETERS
~~~~~~~~~~

shortreads SHORTREADS
```````````````````````

should bowtie alignments be done using paramter for very short reads (~20 bp). default = TRUE


STAGE 4 PARAMETERS
~~~~~~~~~~

MACS_pvalue MACS_PVALUE
```````````````````````

pvalue cutoff for peak calling in MACS2. default = 0.00001 

peakslop PEAKSLOP
```````````````````````

Number of basespairs to extend peaks on both sides. default = 1000


peakinput PEAKINPUT
```````````````````````

Name of user supplied peaks file.  If NULL Mango will use peaks determined from MACS2 analysis. default = NULL


STAGE 5 PARAMETERS
~~~~~~~~~~

distcutrangemin DISTCUTRANGEMIN
```````````````````````

When Mango determines the self-ligation cutoff this is the minimum distance it will consider. default = 1000


distcutrangemax DISTCUTRANGEAX
```````````````````````

When Mango determines the self-ligation cutoff this is the maximum distance it will consider. default = 100000


biascut BIASCUT
```````````````````````

Mango exlcudes very short distance PETS since they tend to arise from self-ligation of a single DNA framgent as opposed to interligation of two interacting fragments. To determine this distnce cutoff Mango determines the fraction of PETs at each distance that come from self-ligation and sets the cutoff at the point where the fraction is less than or equal to BIASCUT. default = 0.05
    
maxPval MAXPVAL
```````````````````````

P-value cutoff for significant interactions. default = 0.01

numofbins NUMOFBINS
```````````````````````

number of bins to use for binomial p-value calculations. default = 30
    
corrMethod CORRMETHOD
```````````````````````

Method to use for correction of mulitply hypothesis testing.  See (http://stat.ethz.ch/R-manual/R-devel/library/stats/html/p.adjust.html) for more details. default = BY
    
maxinteractingdist MAXINTERACTINGDIST
```````````````````````

The maximum disance (in basepairs) considered for interaction. default = 10000000
    
FDR FDR
```````````````````````

FDR cutoff for interactions. default = 0.01
    
minPETS MINPETS
```````````````````````

The minimum number of PETs required for an interaction (applied after FDR filtering). default = 2

reportallpairs REPORTALLPAIRS
```````````````````````

Should all pairs be reported or just significant pairs (TRUE or FALSE). default = FALSE
